x/vote/keeper/msg_server.go

```diff
@@ -56,21 +51,27 @@ func (s msgServer) Vote(c context.Context, req *types.VoteRequest) (*types.VoteR
 	case poll.Is(vote.Pending):
 		event = event.AppendAttributes(sdk.NewAttribute(types.AttributeKeyPollState, vote.Pending.String()))
 
-		return &types.VoteResponse{Log: fmt.Sprintf("not enough votes to confirm poll %s yet", poll.GetKey())}, nil
+		return &types.VoteResponse{Log: fmt.Sprintf("not enough votes to confirm poll %s yet", poll.GetID().String())}, nil
 	case poll.Is(vote.Failed):
 		event = event.AppendAttributes(sdk.NewAttribute(types.AttributeKeyPollState, vote.Failed.String()))
 
-		return &types.VoteResponse{Log: fmt.Sprintf("poll %s failed", poll.GetKey())}, nil
+		return &types.VoteResponse{Log: fmt.Sprintf("poll %s failed", poll.GetID().String())}, nil
 	case poll.Is(vote.Expired):
-		return &types.VoteResponse{Log: fmt.Sprintf("poll %s expired", poll.GetKey())}, nil
+		return &types.VoteResponse{Log: fmt.Sprintf("poll %s expired", poll.GetID().String())}, nil
 	case result != nil:
 		_, ok := result.(*vote.Vote)
 		if !ok {
-			return nil, fmt.Errorf("result of poll %s has wrong type, expected *exported.Vote, got %T", poll.GetKey().String(), poll.GetResult())
+			return nil, fmt.Errorf("result of poll %s has wrong type, expected *exported.Vote, got %T", poll.GetID().String(), poll.GetResult())
+		}
+
+		pollModuleMetadata := poll.GetModuleMetadata()
+		voteHandler := s.GetVoteRouter().GetHandler(pollModuleMetadata.Module)
+		if voteHandler == nil {
+			return nil, fmt.Errorf("unknown module for vote %s", pollModuleMetadata.Module)
 		}
 
 		if err := voteHandler.HandleResult(ctx, result); err != nil {
-			return &types.VoteResponse{Log: fmt.Sprintf("vote handler failed %s", err.Error())}, nil
+			return nil, err
```

Comment:
Based on discussion with @sammy1991106, to avoid this situation, the vote msg server will pass in a cached context to the vote handler, and revert it if the handler returns an error. If it's successful, it'll commit it. This way the validator is not penalized. And individual modules don't need to worry about simultaneous polls of the same tx.